replace findall with subset

yakir12 opened the pull request:
Using `DataFrames.subset` seems a more appropriate to select a subset of signals from a single recording.

yakir12 commented:
Actually, I see a few places where a filter could also be replaced with a subset, but to be honest the improvement would be marginal. On the other hand, if you were ok with using `DataFramesMeta` or `DataFrameMacros` on top of DataFrames then things could get prettier indeed:
from
```julia
filter(s -> length(s.channels) > 1 && duration(s.span) > Minute(5), signals)
```
to (in the case of `DataFramesMeta`)
```julia
@rsubset(signals, length(:channels) > 1, duration(:span) > Minute(5))
```
or because it's a macro (not sure what the policy is about stuff like that, I personally am not a fan of removing the commas)
```julia
@rsubset signals length(:channels) > 1 duration(:span) > Minute(5)
```

The advantage would be cleaner code in all the examples.

ararslan reviewed (commented):
LGTM pending one suggestion, nice catch!

ararslan commented:
> the improvement would be marginal

Given that `filter` exists to satisfy the collections API but `subset` is the [officially recommended way](https://dataframes.juliadata.org/stable/man/working_with_dataframes/#Subsetting-functions) I think it'd be a worthwhile change.

> if you were ok with using DataFramesMeta or DataFrameMacros on top of DataFrames

Personally I think it's worthwhile to keep things simple in the tour and just use DataFrames directly; using a layer of abstraction atop DataFrames I think just adds more overhead for newcomers to learn. That's my $0.02 anyway.

yakir12 commented:
OK! I'll try to clean up some of those things, use YAS, and keep it to vanilla DataFrames.

jrevels reviewed (changes requested):
only skimmed (will lean on @ararslan for dedicated review) but ❤️ this, `subset` is way cleaner IMO

only thing I called out was a style nit :)

yakir12 commented:
I reverted the `.=` syntax for new columns in `DataFrames` since it depends on newer Julia versions than 1.4 (so now all tests pass). In an example code like this I hardly think it matters. I also reverted the auto-formatting of that for loop as well. Let me know what else is left :slightly_smiling_face:

ararslan commented on examples/tour.jl:
```diff
@@ -130,35 +133,38 @@ combine(groupby(signals, :recording), nrow)
 target = rand(signals.recording)
-signals_copy = copy(signals) # we're gonna keep using `signals` afterwards, so let's work with a copy
-filter!(s -> s.recording != target #=|| (rm(s.file_path); false)=#, signals_copy)
+grps = groupby(signals, :recording)
+# foreach(rm, grps[(target,)].file_path)
+keep = DataFrame(grps[Not((target,))])
```
```suggestion
subset!(signals_copy, [:recording, :file_path] => ByRow() do rec, path
            if rec == target
                #rm(path)  # Uncomment this line to actually delete the sample data
                return false
            end
            return true
        end)
```
This seems a bit more natural to me; I had to stare at the `groupby` + inverted index approach a bit before realizing how it worked.

ararslan commented on examples/tour.jl:
```diff
@@ -130,35 +133,38 @@ combine(groupby(signals, :recording), nrow)
-sig = first(sig for sig in eachrow(signals) if any(within_signal(ann, sig) for ann in eachrow(annotations)))
-transform(filter(ann -> within_signal(ann, sig), annotations),
-          :span => ByRow(span -> load(sig, translate(span, -start(sig.span)))) => :samples)
+signals.signal = 1:nrow(signals) # define an ID for each signal (i.e. row)
+df = innerjoin(annotations, signals; on=:recording, makeunique=true) # join signals and annotations on recording
+subset!(df, [:span_1, :span] => ByRow(TimeSpans.contains)) # keep only the instances where the annotation's span is contained by the signal's span
+grp = first(groupby(df, :signal)) # arbitrarily pick the first signal
+grp.samples = load.(eachrow(grp), translate.(grp.span, -start.(grp.span)))
```
```suggestion
annotated_signals = innerjoin(signals, select(annotations, :recording, :id,
                                              :span => :annotated_span);
                              on=:recording)
subset!(annotated_signals, [:span, :annotated_span] => ByRow(TimeSpans.contains))
annotated_signals = combine(groupby(annotated_signals, :recording), first)
transform(annotated_signals, AsTable(:) => ByRow() do row
              return load(row, translate(row.annotated_span, -start(row.span)))
          end => :samples)
```
I had to think pretty hard about the `makeunique` behavior then match my assumption of which one got assigned the suffix to the order they were passed to `subset!`. Also the signals already have an identifier: the recording ID. (IIRC signals table rows are technically only unique on `recording`, `kind`, and `span` but I think this is reasonable for the example and for most practical uses.)

ararslan commented on examples/tour.jl:
```diff
@@ -167,9 +173,8 @@ transform(filter(ann -> within_signal(ann, sig), annotations),
-samples = load(sig)
-transform(filter(ann -> within_signal(ann, sig), annotations),
-          :span => ByRow(span -> view(samples, :, translate(span, -start(sig.span)))) => :samples)
+samples = load(grp[1, :])
+transform!(grp, :span => ByRow(s -> view(samples, :, translate(s, -start(s)))) => :samples)
```
```suggestion
samples = load(first(annotated_signals))
transform(annotated_signals, :span => ByRow(s -> view(samples, :, translate(s, -start(s)))) => :samples)
```
This suggestion is paired with the previous

ararslan commented:
> it depends on newer Julia versions than 1.4 (so now all tests pass).

Doesn't have to be part of this PR but we could surely drop support for Julia < 1.6 in this package since that's the oldest officially supported release upstream.

yakir12 commented:
Two remaining points:
1. I gave you some push-back on one of your points there, let me know what you think. 
2. codecov complains that there is a 0.01% regression on coverage, but when I go look to see what it's talking about, it claims that:
![image](https://user-images.githubusercontent.com/3281957/174246513-49004f2a-d858-4909-b884-b62f44703d84.png)

What does a 0.01% regression mean anyways...? Silly.

jrevels commented:
closing as stale (but can accept if reopened/updated against `main` + comments addressed)
